### PR TITLE
[SPARK-49445][UI] Support show tooltip in the progress bar of UI

### PR DESCRIPTION
**Pull request by cxzl25**
### What changes were proposed in this pull request?
This PR aims to support show tooltip in the progress bar of UI.

### Why are the changes needed?
Currently, too much content has been added to the progress bar content in the UI, but the width of the progress bar is limited and it cannot display so much content.

image

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
local test

image

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by yaooqinn** [comment]
Can we only add tips for failed ones?

**Comment by cxzl25** [comment]
> Can we only add tips for failed ones?

Because now the Stage of the progress bar won't show the running Task as long as there is a failed task, that's why I want to add the Running Task to the tips, so that it's easy to view.

https://github.com/apache/spark/blob/ebb1975ae4c16da16c61b4a58159d7e7da65f1d2/core/src/main/scala/org/apache/spark/ui/UIUtils.scala#L452

**Comment by yaooqinn** [comment]
We now perform twice for the contents of this spanning element, will it be a risk for perf-regression？

**Comment by cxzl25** [comment]
> will it be a risk for perf-regression

Changed to avoid concatenating and sorting a larger string.

File: core/src/main/scala/org/apache/spark/ui/UIUtils.scala [hunk 1]
```diff
@@ -446,16 +446,24 @@ private[spark] object UIUtils extends Logging {
     val startRatio = if (total == 0) 0.0 else (boundedStarted.toDouble / total) * 100
     val startWidth = "width: %s%%".format(startRatio)
 
+    val killTaskReasonText = reasonToNumKilled.toSeq.sortBy(-_._2).map {
+        case (reason, count) => s" ($count killed: $reason)"
+      }.mkString
+    val progressTitle = s"$completed / $total" + {
```
Comment (yaooqinn) [inline core/src/main/scala/org/apache/spark/ui/UIUtils.scala:452]:
```suggestion
    val progressTitle = s"$completed/$total" + {
```

File: core/src/main/scala/org/apache/spark/ui/UIUtils.scala [hunk 2]
```diff
@@ -446,16 +446,24 @@ private[spark] object UIUtils extends Logging {
...
+      <span style="text-align:center; position:absolute; width:100%;" title={progressTitle}>
         {completed}/{total}
         { if (failed == 0 && skipped == 0 && started > 0) s"($started running)" }
         { if (failed > 0) s"($failed failed)" }
         { if (skipped > 0) s"($skipped skipped)" }
-        { reasonToNumKilled.toSeq.sortBy(-_._2).map {
-            case (reason, count) => s"($count killed: $reason)"
-          }
-        }
+        { killTaskReasonText }
```
Comment (yaooqinn) [inline core/src/main/scala/org/apache/spark/ui/UIUtils.scala:466]: Can it be replaced with {progressTitle}?

Comment (cxzl25) [reply core/src/main/scala/org/apache/spark/ui/UIUtils.scala:466]: The logic of running tasks displayed by the two is different.

```
if (started > 0) s" ($started running)" else ""
```

---

```
{ if (failed == 0 && skipped == 0 && started > 0) s"($started running)" } 
```

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
Merged to master. Thank you

**Pull request closed without merging**